compiler/rustc_mir/src/transform/mod.rs

```diff
@@ -318,6 +321,72 @@ fn mir_promoted(
     (tcx.alloc_steal_mir(body), tcx.alloc_steal_promoted(promoted))
 }
 
+/// Compute the MIR that is used during CTFE (and thus has no optimizations run on it)
+fn mir_for_ctfe<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> &'tcx Body<'tcx> {
+    let did = def_id.expect_local();
+    if let Some(def) = ty::WithOptConstParam::try_lookup(did, tcx) {
+        tcx.mir_for_ctfe_of_const_arg(def)
+    } else {
+        tcx.arena.alloc(inner_mir_for_ctfe(tcx, ty::WithOptConstParam::unknown(did)))
+    }
+}
+
+/// Same as `mir_for_ctfe`, but used to get the MIR of a const generic parameter.
+/// The docs on `WithOptConstParam` explain this a bit more, but the TLDR is that
+/// we'd get cycle errors with `mir_for_ctfe`, because typeck would need to typeck
+/// the const parameter while type checking the main body, which in turn would try
+/// to type check the main body again.
+fn mir_for_ctfe_of_const_arg<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    (did, param_did): (LocalDefId, DefId),
+) -> &'tcx Body<'tcx> {
+    tcx.arena.alloc(inner_mir_for_ctfe(
+        tcx,
+        ty::WithOptConstParam { did, const_param_did: Some(param_did) },
+    ))
+}
+
+fn inner_mir_for_ctfe(tcx: TyCtxt<'_>, def: ty::WithOptConstParam<LocalDefId>) -> Body<'_> {
+    // FIXME: don't duplicate this between the optimized_mir/mir_for_ctfe queries
+    if tcx.is_constructor(def.did.to_def_id()) {
+        // There's no reason to run all of the MIR passes on constructors when
+        // we can just output the MIR we want directly. This also saves const
+        // qualification and borrow checking the trouble of special casing
+        // constructors.
+        return shim::build_adt_ctor(tcx, def.did.to_def_id());
+    }
+
+    assert_ne!(
+        tcx.hir().body_const_context(def.did),
+        None,
+        "mir_for_ctfe should not be used for runtime functions"
+    );
+
+    let mut body = tcx.mir_drops_elaborated_and_const_checked(def).borrow().clone();
+
+    #[rustfmt::skip]
+    let optimizations: &[&dyn MirPass<'_>] = &[
+        &const_prop::ConstProp,
```

Comment:
This part of my earlier (marked-as-resolved) comment is not yet resolved:

> Please explain why we still need to run this. Doesn't this lead to duplicate lints (and duplicate work!) for const fn as we run the same pass twice on them?

Reply:
I guess if a `const fn` is not used at runtime and no metadata is encoded for the chosen crate type, `optimized_mir` will never be invoked.

Reply:
What bjorn3 said.

There is no duplication of emitted lints, since we deduplicate lints already, and it would be exact duplicates.

Reply:
But the optimization pass will still run twice for most functions, which does not seem great.

Reply:
I guess we could leave it out for the `mir_for_ctfe`, as CTFE will catch the same errors, but this will cause enormous ui test churn, so I think we should do it in a PR of its own.

Reply:
If the lints are all duplicates, why will it cause ui test churn?

Reply:
If a non-public `const fn` is only used for CTFE, but we don't run const prop on it, then, since `optimized_mir` will never be executed, we won't run const prop at all, so we won't get any lints, but the CTFE run will catch all situations in the actually executed paths. These will then occur as CTFE errors and not const prop lints. So any ui test that has const prop lints on `const fn` only used in CTFE will now not emit those lints anymore, but instead emit CTFE errors

Reply:
I just played with this. Tow more things I noticed there:

* the only way to detect some failures in associated constants is to run const prop. We have tests for this. I don't know how useful this is, but it requires updating some tests
* we run const prop on normal constants, too, but again, not sure how useful that is, I mostly kept it when implementing the early version of const prop in order to keep the change small. We could just say we stop emitting const prop lints in constants (and statics) since they will get evaluated anyway and then we will see CTFE failures only in the actually executed paths.

Reply:
Ok, just turning this off for `const fn` works out fine, since almost all cases in our tests actually generate `optimized_mir` (see the latest commit). Changing anything for const items (associated or not), touches around 10 tests

Reply:
Const items don't have the "other" query run, I assume, so there's no duplication of work there.

Reply:
Running this only on `const` items makes sense, thanks!